model/rhc_connection.go

```diff
@@ -0,0 +1,67 @@
+package model
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/RedHatInsights/sources-api-go/util"
+	"gorm.io/datatypes"
+)
+
+type RhcConnection struct {
+	ID    int64          `gorm:"primaryKey" json:"id"`
+	RhcId string         `json:"rhc_id"`
+	Extra datatypes.JSON `json:"extra,omitempty"`
+	AvailabilityStatus
+	AvailabilityStatusError string    `json:"availability_status_error,omitempty"`
+	CreatedAt               time.Time `json:"created_at"`
+	UpdatedAt               time.Time `json:"updated_at"`
+
+	Sources []Source `gorm:"many2many:source_rhc_connections"`
+}
+
+func (r *RhcConnection) UpdateFromRequest(input *RhcConnectionUpdateRequest) {
+	r.Extra = input.Extra
+}
+
+func (r *RhcConnection) ToEvent() interface{} {
+	id := strconv.FormatInt(r.ID, 10)
+
+	asEvent := AvailabilityStatusEvent{
+		AvailabilityStatus: util.StringValueOrNil(r.AvailabilityStatus.AvailabilityStatus),
+		LastAvailableAt:    util.DateTimeToRecordFormat(r.LastAvailableAt),
+		LastCheckedAt:      util.DateTimeToRecordFormat(r.LastCheckedAt),
+	}
+
+	rhcConnectionEvent := &RhcConnectionEvent{
+		ID:                      &id,
+		RhcId:                   &r.RhcId,
+		Extra:                   r.Extra,
+		AvailabilityStatusEvent: asEvent,
+		AvailabilityStatusError: &r.AvailabilityStatusError,
+		CreatedAt:               util.DateTimeToRecordFormat(r.CreatedAt),
+		UpdatedAt:               util.DateTimeToRecordFormat(r.UpdatedAt),
+	}
+
+	return rhcConnectionEvent
+}
+
+func (r *RhcConnection) ToResponse() *RhcConnectionResponse {
+	id := strconv.FormatInt(r.ID, 10)
+
+	var sourceIds []string
+	if len(r.Sources) > 0 {
+		for _, src := range r.Sources {
+			sourceIds = append(sourceIds, strconv.FormatInt(src.ID, 10))
+		}
+	}
```

Comment:
For this I'd do something like:

```suggestion
	sourceIds := make([]string, len(r.Sources))
	for i, src := range r.Sources {
		sourceIds[i] = strconv.FormatInt(src.ID, 10)
	}
```

1. Allocating exactly the size you need in the array!
2. Looping over without checking bounds (the loop wouldn't run unless there is a source, and there is almost guaranteed there will be one)

Reply:
Oooof yes. I confused the `len` and `cap` arguments in the `make` call, and I was being hit by some weird "index out of bounds" errors. I was doing the following:

`make([]string, 0, len(r.Sources)`

Thus allocating a zero length array. After that, by trying to access it with the `sourceIds[i]` statement I was having that error. That's why I was using a more "manual" loop.

I have refactored it now.